File: src/za/co/adhd_developers/questions/Question43.java
package za.co.adhd_developers.questions;

import za.co.adhd_developers.tools.Question;
import za.co.adhd_developers.tools.Utils;

import java.util.ArrayList;
import java.util.List;

public class Question43 implements Question
{
    private long answer = 0L;
    @Override
    public void doWork()
    {
        List<String> permutations = new ArrayList<>();

        Utils.permutation("0123456789", permutations);

        for (String permutation : permutations)
        {
            if (permutation.startsWith("0"))
            {
                continue;
            }
            if (!Utils.isPandigital(permutation, 0, 9))
            {
                continue;
            }
            if (!Utils.hasSubStringDivisibility(permutation))
            {
                continue;
            }
            this.answer += Long.valueOf(permutation);
        }
    }

    @Override
    public void printAnswer()
    {
        System.out.println("=================================");
        System.out.println("Question 43");
        System.out.println("The sum of all 0 to 9 pandigital numbers with the sub-string divisibility property, is: " + this.answer);
        System.out.println("=================================");
    }
}

File: src/za/co/adhd_developers/kotlin/questions/Question43.kt
package za.co.adhd_developers.kotlin.questions

import za.co.adhd_developers.kotlin.Question
import za.co.adhd_developers.kotlin.tools.KPermUtil
import za.co.adhd_developers.kotlin.tools.isPandigital

class Question43 : Question {

    var answer = 0L

    override fun doWork() {
        val numbers = "1234567890"

        val permUtil = KPermUtil(numbers.map { c -> c.toString() }.toTypedArray())

        var currentPerm = permUtil.next()

        while (currentPerm != null) {

            val possible = currentPerm.joinToString("")

            if (possible.startsWith("0")) {
                currentPerm = permUtil.next()
                continue
            }

            if (!isPandigital(possible, 0)) {
                currentPerm = permUtil.next()
                continue
            }

            var index = 1
            if (possible.substring(index,index+3).toInt().rem(2) != 0) {
                currentPerm = permUtil.next()
                continue
            }
            index++
            if (possible.substring(index,index+3).toInt().rem(3) != 0) {
                currentPerm = permUtil.next()
                continue
            }
            index++
            if (possible.substring(index,index+3).toInt().rem(5) != 0) {
                currentPerm = permUtil.next()
                continue
            }
            index++
            if (possible.substring(index,index+3).toInt().rem(7) != 0) {
                currentPerm = permUtil.next()
                continue
            }
            index++
            if (possible.substring(index,index+3).toInt().rem(11) != 0) {
                currentPerm = permUtil.next()
                continue
            }
            index++
            if (possible.substring(index,index+3).toInt().rem(13) != 0) {
                currentPerm = permUtil.next()
                continue
            }
            index++
            if (possible.substring(index,index+3).toInt().rem(17) != 0) {
                currentPerm = permUtil.next()
                continue
            }

            answer = answer.plus(possible.toLong())
            currentPerm = permUtil.next()
        }
    }

    override fun printAnswer() {
        println("=================================")
        println("Question 43")
        println("The sum of all 0 to 9 pandigital numbers with the sub-string divisibility property, is: " + this.answer)
        println("=================================")
    }
}